iothub/device/src/Authentication/ClientAuthenticationWithTokenRefresh.cs

```diff
@@ -2,48 +2,134 @@
             string moduleId = default,
             TimeSpan suggestedTimeToLive = default,
             int timeBufferPercentage = default)
-            : base(suggestedTimeToLive,
-                  timeBufferPercentage)
         {
             Argument.AssertNotNullOrWhiteSpace(deviceId, nameof(deviceId));
             DeviceId = deviceId;
 
+            // The module Id is optional so we only check whether it is whitespace or not here.
             if (moduleId != null
                 && moduleId.IsNullOrWhiteSpace())
             {
                 throw new ArgumentException("Module Id cannot be white space.");
             }
+
             ModuleId = moduleId;
+
+            if (suggestedTimeToLive < TimeSpan.Zero)
+            {
+                throw new ArgumentOutOfRangeException(nameof(suggestedTimeToLive));
+            }
+
+            if (timeBufferPercentage < 0 || timeBufferPercentage > 100)
+            {
+                throw new ArgumentOutOfRangeException(nameof(timeBufferPercentage));
+            }
+
+            SuggestedTimeToLive = suggestedTimeToLive == default
+                ? s_defaultSasTimeToLive
+                : suggestedTimeToLive;
+
+            TimeBufferPercentage = timeBufferPercentage == default
+                ? DefaultSasRenewalBufferPercentage
+                : timeBufferPercentage;
+
+            ExpiresOn = DateTime.UtcNow.AddSeconds(-SuggestedTimeToLive.TotalSeconds);
+            Debug.Assert(IsExpiring);
+            UpdateTimeBufferSeconds(SuggestedTimeToLive.TotalSeconds);
+        }
+
+        /// <summary>
+        /// Creates an instance of this class.
+        /// </summary>
+        /// <param name="connectionString">
+        /// The connection string containing the device Id, optional module Id, shared access key name
+        /// and shared access key to be used for authenticating with IoT hub service.
+        /// </param>
+        /// <param name="suggestedTimeToLive">
+        /// The suggested time to live value for the generated SAS tokens.
+        /// The default value is 1 hour.
+        /// </param>
+        /// <param name="timeBufferPercentage">
+        /// The time buffer before expiry when the token should be renewed, expressed as a percentage of the time to live.
+        /// The default behavior is that the token will be renewed when it has 15% or less of its lifespan left.
+        /// </param>
+        /// <exception cref="ArgumentNullException">Thrown when <paramref name="connectionString"/> is null.</exception>
+        /// <exception cref="ArgumentException">Thrown when <paramref name="connectionString"/> is empty or whitespace.</exception>
+        /// <exception cref="ArgumentOutOfRangeException">
+        /// Thrown if <paramref name="suggestedTimeToLive"/> is a negative timespan, or if <paramref name="timeBufferPercentage"/> is outside the range 0-100.
+        /// </exception>
+        public ClientAuthenticationWithTokenRefresh(
```

Comment:
We can extract out the device Id and module Id in 
```c#
        public ClientAuthenticationWithSharedAccessKeyRefresh(
            string connectionString,
            TimeSpan sasTokenTimeToLive = default,
            int sasTokenRenewalBuffer = default)
            : base(
                  connectionString,
                  sasTokenTimeToLive,
                  sasTokenRenewalBuffer)
        {
            Argument.AssertNotNullOrWhiteSpace(connectionString, nameof(connectionString));

            _iotHubConnectionString = IotHubConnectionStringParser.Parse(connectionString);
            SharedAccessKey = _iotHubConnectionString.SharedAccessKey;
            SharedAccessKeyName = _iotHubConnectionString.SharedAccessKeyName;
        }
```

and use that to call the first constructor.

That way, we won't need to parse out the connection string in here again, and we can get rid of this constructor.